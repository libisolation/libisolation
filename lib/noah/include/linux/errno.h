#ifndef NOAH_LINUX_ERRNO_H
#define NOAH_LINUX_ERRNO_H

#include <errno.h>
#include "linux/const.h"

#define LINUX_ERRNO(_) \
  DECL_LINUX(_,EPERM,           1)                                  \
  DECL_LINUX(_,ENOENT,          2)                                  \
  DECL_LINUX(_,ESRCH,           3)                                  \
  DECL_LINUX(_,EINTR,           4)                                  \
  DECL_LINUX(_,EIO,             5)                                  \
  DECL_LINUX(_,ENXIO,           6)                                  \
  DECL_LINUX(_,E2BIG,           7)                                  \
  DECL_LINUX(_,ENOEXEC,         8)                                  \
  DECL_LINUX(_,EBADF,           9)                                  \
  DECL_LINUX(_,ECHILD,          10)                                 \
  DECL_LINUX(_,EAGAIN,          11)                                 \
  DECL_LINUX(_,ENOMEM,          12)                                 \
  DECL_LINUX(_,EACCES,          13)                                 \
  DECL_LINUX(_,EFAULT,          14)                                 \
  DECL_LINUX(_,ENOTBLK,         15)                                 \
  DECL_LINUX(_,EBUSY,           16)                                 \
  DECL_LINUX(_,EEXIST,          17)                                 \
  DECL_LINUX(_,EXDEV,           18)                                 \
  DECL_LINUX(_,ENODEV,          19)                                 \
  DECL_LINUX(_,ENOTDIR,         20)                                 \
  DECL_LINUX(_,EISDIR,          21)                                 \
  DECL_LINUX(_,EINVAL,          22)                                 \
  DECL_LINUX(_,ENFILE,          23)                                 \
  DECL_LINUX(_,EMFILE,          24)                                 \
  DECL_LINUX(_,ENOTTY,          25)                                 \
  DECL_LINUX(_,ETXTBSY,         26)                                 \
  DECL_LINUX(_,EFBIG,           27)                                 \
  DECL_LINUX(_,ENOSPC,          28)                                 \
  DECL_LINUX(_,ESPIPE,          29)                                 \
  DECL_LINUX(_,EROFS,           30)                                 \
  DECL_LINUX(_,EMLINK,          31)                                 \
  DECL_LINUX(_,EPIPE,           32)                                 \
  DECL_LINUX(_,EDOM,            33)                                 \
  DECL_LINUX(_,ERANGE,          34)                                 \
  DECL_LINUX(_,EDEADLK,         35)                                 \
  DECL_LINUX(_,ENAMETOOLONG,    36)                                 \
  DECL_LINUX(_,ENOLCK,          37)                                 \
  DECL_LINUX(_,ENOSYS,          38)                                 \
  DECL_LINUX(_,ENOTEMPTY,       39)                                 \
  DECL_LINUX(_,ELOOP,           40)                                 \
  DECL_ALIAS(_,EWOULDBLOCK,     LINUX_EAGAIN)                       \
  DECL_LINUX(_,ENOMSG,          42)                                 \
  DECL_LINUX(_,EIDRM,           43)                                 \
  DECL_LINUX(_,ECHRNG,          44,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EL2NSYNC,        45,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EL3HLT,          46,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EL3RST,          47,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ELNRNG,          48,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EUNATCH,         49,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ENOCSI,          50,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EL2HLT,          51,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EBADE,           52,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EBADR,           53,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EXFULL,          54,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ENOANO,          55,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EBADRQC,         56,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EBADSLT,         57,               LINUX_SPECIFIC)   \
  DECL_ALIAS(_,EDEADLOCK,       LINUX_EDEADLK)                      \
  DECL_LINUX(_,EBFONT,          59,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ENOSTR,          60)                                 \
  DECL_LINUX(_,ENODATA,         61)                                 \
  DECL_LINUX(_,ETIME,           62)                                 \
  DECL_LINUX(_,ENOSR,           63)                                 \
  DECL_LINUX(_,ENONET,          64,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ENOPKG,          65,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EREMOTE,         66)                                 \
  DECL_LINUX(_,ENOLINK,         67)                                 \
  DECL_LINUX(_,EADV,            68,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ESRMNT,          69,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ECOMM,           70,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EPROTO,          71)                                 \
  DECL_LINUX(_,EMULTIHOP,       72)                                 \
  DECL_LINUX(_,EDOTDOT,         73,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EBADMSG,         74)                                 \
  DECL_LINUX(_,EOVERFLOW,       75)                                 \
  DECL_LINUX(_,ENOTUNIQ,        76,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EBADFD,          77,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EREMCHG,         78,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ELIBACC,         79,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ELIBBAD,         80,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ELIBSCN,         81,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ELIBMAX,         82,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ELIBEXEC,        83,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EILSEQ,          84)                                 \
  DECL_LINUX(_,ERESTART,        85,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,ESTRPIPE,        86,               LINUX_SPECIFIC)   \
  DECL_LINUX(_,EUSERS,          87)                                 \
  DECL_LINUX(_,ENOTSOCK,        88)                                 \
  DECL_LINUX(_,EDESTADDRREQ,    89)                                 \
  DECL_LINUX(_,EMSGSIZE,        90)                                 \
  DECL_LINUX(_,EPROTOTYPE,      91)                                 \
  DECL_LINUX(_,ENOPROTOOPT,     92)                                 \
  DECL_LINUX(_,EPROTONOSUPPORT, 93)                                 \
  DECL_LINUX(_,ESOCKTNOSUPPORT, 94)                                 \
  DECL_LINUX(_,EOPNOTSUPP,      95)                                 \
  DECL_ALIAS(_,ENOTSUP,         LINUX_EOPNOTSUPP)                   \
  DECL_LINUX(_,EPFNOSUPPORT,    96)                                 \
  DECL_LINUX(_,EAFNOSUPPORT,    97)                                 \
  DECL_LINUX(_,EADDRINUSE,      98)                                 \
  DECL_LINUX(_,EADDRNOTAVAIL,   99)                                 \
  DECL_LINUX(_,ENETDOWN,        100)                                \
  DECL_LINUX(_,ENETUNREACH,     101)                                \
  DECL_LINUX(_,ENETRESET,       102)                                \
  DECL_LINUX(_,ECONNABORTED,    103)                                \
  DECL_LINUX(_,ECONNRESET,      104)                                \
  DECL_LINUX(_,ENOBUFS,         105)                                \
  DECL_LINUX(_,EISCONN,         106)                                \
  DECL_LINUX(_,ENOTCONN,        107)                                \
  DECL_LINUX(_,ESHUTDOWN,       108)                                \
  DECL_LINUX(_,ETOOMANYREFS,    109)                                \
  DECL_LINUX(_,ETIMEDOUT,       110)                                \
  DECL_LINUX(_,ECONNREFUSED,    111)                                \
  DECL_LINUX(_,EHOSTDOWN,       112)                                \
  DECL_LINUX(_,EHOSTUNREACH,    113)                                \
  DECL_LINUX(_,EALREADY,        114)                                \
  DECL_LINUX(_,EINPROGRESS,     115)                                \
  DECL_LINUX(_,ESTALE,          116)                                \
  DECL_LINUX(_,EUCLEAN,         117,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,ENOTNAM,         118,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,ENAVAIL,         119,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EISNAM,          120,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EREMOTEIO,       121,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EDQUOT,          122)                                \
  DECL_LINUX(_,ENOMEDIUM,       123,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EMEDIUMTYPE,     124,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,ECANCELED,       125)                                \
  DECL_LINUX(_,ENOKEY,          126,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EKEYEXPIRED,     127,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EKEYREVOKED,     128,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EKEYREJECTED,    129,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EOWNERDEAD,      130)                                \
  DECL_LINUX(_,ENOTRECOVERABLE, 131)                                \
  DECL_LINUX(_,ERFKILL,         132,              LINUX_SPECIFIC)   \
  DECL_LINUX(_,EHWPOISON,       133,              LINUX_SPECIFIC)   \
  FROM_DARWN(_,EPERM,           EFTYPE)                             \
  FROM_DARWN(_,EPERM,           EAUTH)                              \
  FROM_DARWN(_,EPERM,           ENEEDAUTH)                          \
  FROM_DARWN(_,ENOEXEC,         EBADEXEC)                           \
  FROM_DARWN(_,ENOEXEC,         EBADARCH)                           \
  FROM_DARWN(_,ENOEXEC,         EBADMACHO)                          \
  FROM_DARWN(_,ENODATA,         ENOATTR)                            \
  FROM_DARWN(_,EPERM,           ENOPOLICY)                          \
  FROM_DARWN(_,EPERM,           EBADRPC)                            \

DECLARE_CENUM(errno, LINUX_ERRNO);
DECLARE_CSTR_FUNC(errno, LINUX_ERRNO);
DECLARE_CMAP_FUNC(darwin_to_linux, errno, LINUX_ERRNO);
DECLARE_CMAP_FUNC(linux_to_darwin, errno, LINUX_ERRNO);

#endif
