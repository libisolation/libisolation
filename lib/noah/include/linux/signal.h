/*-
 * Copyright (c) 2016 Yuichi Nishiwaki, Takaya Saeki
 * Copyright (c) 2015 Dmitry Chagin
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $FreeBSD$
 */

#ifndef NOAH_LINUX_SIGNAL_H
#define NOAH_LINUX_SIGNAL_H

#include "linux/common.h"
#include "linux/const.h"
#include "types.h"
#include <signal.h>

/* signaling */
#define LINUX_STSIGNUM(_)\
	DECL_LINUX(_,SIGHUP,		1)\
	DECL_LINUX(_,SIGINT,		2)\
	DECL_LINUX(_,SIGQUIT,		3)\
	DECL_LINUX(_,SIGILL,		4)\
	DECL_LINUX(_,SIGTRAP,		5)\
	DECL_LINUX(_,SIGABRT,		6)\
	DECL_ALIAS(_,SIGIOT,		LINUX_SIGABRT)\
	DECL_LINUX(_,SIGBUS,		7)\
	DECL_LINUX(_,SIGFPE,		8)\
	DECL_LINUX(_,SIGKILL,		9)\
	DECL_LINUX(_,SIGUSR1,		10)\
	DECL_LINUX(_,SIGSEGV,		11)\
	DECL_LINUX(_,SIGUSR2,		12)\
	DECL_LINUX(_,SIGPIPE,		13)\
	DECL_LINUX(_,SIGALRM,		14)\
	DECL_LINUX(_,SIGTERM,		15)\
	DECL_LINUX(_,SIGSTKFLT,		16, LINUX_SPECIFIC)/* On x86, SIGSTKFLT is never used */\
	DECL_LINUX(_,SIGCHLD,		17)\
	DECL_LINUX(_,SIGCONT,		18)\
	DECL_LINUX(_,SIGSTOP,		19)\
	DECL_LINUX(_,SIGTSTP,		20)\
	DECL_LINUX(_,SIGTTIN,		21)\
	DECL_LINUX(_,SIGTTOU,		22)\
	DECL_LINUX(_,SIGURG,		23)\
	DECL_LINUX(_,SIGXCPU,		24)\
	DECL_LINUX(_,SIGXFSZ,		25)\
	DECL_LINUX(_,SIGVTALRM,		26)\
	DECL_LINUX(_,SIGPROF,		27)\
	DECL_LINUX(_,SIGWINCH,		28)\
	DECL_LINUX(_,SIGIO,		29)\
	DECL_ALIAS(_,SIGPOLL,		LINUX_SIGIO)\
	DECL_LINUX(_,SIGPWR,		30, LINUX_SPECIFIC) /* Noah vkernel wll never send SIGPWR */ \
	DECL_LINUX(_,SIGSYS,		31)\

#define LINUX_SIGTBLSZ 31
#define LINUX_SIGRTMIN 32
#define LINUX_SIGRTMAX 64

DECLARE_CENUM(signum, LINUX_STSIGNUM);
DECLARE_CMAP_FUNC(darwin_to_linux, signum, LINUX_STSIGNUM);
DECLARE_CMAP_FUNC(linux_to_darwin, signum, LINUX_STSIGNUM);
DECLARE_CSTR_FUNC(signum, LINUX_STSIGNUM);


#define LINUX_NSIG              64

/* sigaction flags */
#define	LINUX_SA_NOCLDSTOP	0x00000001
#define	LINUX_SA_NOCLDWAIT	0x00000002
#define	LINUX_SA_SIGINFO	0x00000004
#define	LINUX_SA_RESTORER	0x04000000
#define	LINUX_SA_ONSTACK	0x08000000
#define	LINUX_SA_RESTART	0x10000000
#define	LINUX_SA_INTERRUPT	0x20000000
#define	LINUX_SA_NOMASK		0x40000000
#define	LINUX_SA_ONESHOT	0x80000000

/* sigset */
typedef struct {
  uint64_t	__mask;
} l_sigset_t;

void linux_to_darwin_sigset(l_sigset_t *, sigset_t *);
void darwin_to_linux_sigset(sigset_t *, l_sigset_t *);

/* primitives to manipulate sigset_t */
#define	LINUX_SIGEMPTYSET(setp)		((setp)->__mask = 0)
#define	LINUX_SIGISMEMBER(setp, sig)	(1UL & ((setp)->__mask >> ((sig) - 1)))
#define	LINUX_SIGADDSET(setp, sig)	((setp)->__mask |= (1UL << ((sig) - 1)))
#define	LINUX_SIGDELSET(setp, sig)	((setp)->__mask &= ~(1UL << ((sig) - 1)))

#define	LINUX_SIGSET_ADD(setp, masksetp)	((setp)->__mask |= ((masksetp)->__mask))
#define	LINUX_SIGSET_DEL(setp, masksetp)	((setp)->__mask &= ~((masksetp)->__mask))
#define	LINUX_SIGSET_SET(setp, masksetp)	((setp)->__mask = (masksetp)->__mask)

#define LINUX_SIGSET_TO_UI64(setp)      ((setp)->__mask)

/* sigprocmask actions */
#define	LINUX_SIG_BLOCK		0
#define	LINUX_SIG_UNBLOCK	1
#define	LINUX_SIG_SETMASK	2

/* sigaltstack */
#define	LINUX_MINSIGSTKSZ	2048

typedef gaddr_t	l_handler_t;

typedef struct {
  l_handler_t	lsa_handler;
  l_ulong	lsa_flags;
  l_uintptr_t	lsa_restorer;
  l_sigset_t	lsa_mask;
} l_sigaction_t;

#define LINUX_SIG_DFL ((l_handler_t)0)
#define LINUX_SIG_IGN ((l_handler_t)1)
#define LINUX_SIG_ERR ((l_handler_t)-1)

void linux_to_darwin_sigaction(l_sigaction_t *, struct sigaction *, void *);
void darwin_to_linux_sigaction(struct sigaction *, l_sigaction_t *, gaddr_t);

typedef struct {
  l_uintptr_t	ss_sp;
  l_int		ss_flags;
  l_size_t	ss_size;
} l_stack_t;

#define LINUX_SS_ONSTACK      1
#define LINUX_SS_DISABLE      2
#define LINUX_SS_AUTODISARM   (1U << 31)

struct l_fpstate {
  u_int16_t cwd;
  u_int16_t swd;
  u_int16_t twd;
  u_int16_t fop;
  u_int64_t rip;
  u_int64_t rdp;
  u_int32_t mxcsr;
  u_int32_t mxcsr_mask;
  u_int32_t st_space[32];
  u_int32_t xmm_space[64];
  u_int32_t reserved2[24];
};

struct l_sigcontext {
  l_ulong		sc_r8;
  l_ulong		sc_r9;
  l_ulong		sc_r10;
  l_ulong		sc_r11;
  l_ulong		sc_r12;
  l_ulong		sc_r13;
  l_ulong		sc_r14;
  l_ulong		sc_r15;
  l_ulong		sc_rdi;
  l_ulong		sc_rsi;
  l_ulong		sc_rbp;
  l_ulong		sc_rbx;
  l_ulong		sc_rdx;
  l_ulong		sc_rax;
  l_ulong		sc_rcx;
  l_ulong		sc_rsp;
  l_ulong		sc_rip;
  l_ulong		sc_rflags;
  l_ushort		sc_cs;
  l_ushort		sc_gs;
  l_ushort		sc_fs;
  union {
    l_ushort		sc_ss;
    l_ushort		sc___pad0;
  };
  l_ulong		sc_err;
  l_ulong		sc_trapno;
  l_sigset_t		sc_mask;
  l_ulong		sc_cr2;
  struct l_fpstate	*sc_fpstate;
  l_ulong		sc_reserved1[8];
};

struct l_ucontext {
  l_ulong	uc_flags;
  l_uintptr_t	uc_link;
  l_stack_t	uc_stack;
  struct l_sigcontext	uc_mcontext;
  l_sigset_t	uc_sigmask;
};

#define LINUX_UC_FP_XSTATE	0x1
#define LINUX_UC_SIGCONTEXT_SS	0x2
#define LINUX_UC_STRICT_RESTORE_SS	0x4

#define LINUX_SI_PREAMBLE_SIZE	(4 * sizeof(int))
#define	LINUX_SI_MAX_SIZE	128
#define	LINUX_SI_PAD_SIZE	((LINUX_SI_MAX_SIZE - \
				    LINUX_SI_PREAMBLE_SIZE) / sizeof(l_int))
typedef union l_sigval {
  l_int		sival_int;
  l_uintptr_t	sival_ptr;
} l_sigval_t;

typedef struct l_siginfo {
  l_int		lsi_signo;
  l_int		lsi_errno;
  l_int		lsi_code;
  union {
    l_int	_pad[LINUX_SI_PAD_SIZE];

    struct {
      l_pid_t		_pid;
      l_uid_t		_uid;
    } _kill;

    struct {
      l_timer_t	_tid;
      l_int		_overrun;
      char		_pad[sizeof(l_uid_t) - sizeof(int)];
      union l_sigval	_sigval;
      l_uint		_sys_private;
    } _timer;

    struct {
      l_pid_t		_pid;		/* sender's pid */
      l_uid_t		_uid;		/* sender's uid */
      union l_sigval	_sigval;
    } _rt;

    struct {
      l_pid_t		_pid;		/* which child */
      l_uid_t		_uid;		/* sender's uid */
      l_int		_status;	/* exit code */
      l_clock_t	_utime;
      l_clock_t	_stime;
    } _sigchld;

    struct {
      l_uintptr_t	_addr;	/* Faulting insn/memory ref. */
    } _sigfault;

    struct {
      l_long		_band;	/* POLL_IN,POLL_OUT,POLL_MSG */
      l_int		_fd;
    } _sigpoll;
  } _sifields;
} l_siginfo_t;

#define	lsi_pid		_sifields._kill._pid
#define	lsi_uid		_sifields._kill._uid
#define	lsi_tid		_sifields._timer._tid
#define	lsi_overrun	_sifields._timer._overrun
#define	lsi_sys_private	_sifields._timer._sys_private
#define	lsi_status	_sifields._sigchld._status
#define	lsi_utime	_sifields._sigchld._utime
#define	lsi_stime	_sifields._sigchld._stime
#define	lsi_value	_sifields._rt._sigval
#define	lsi_int		_sifields._rt._sigval.sival_int
#define	lsi_ptr		_sifields._rt._sigval.sival_ptr
#define	lsi_addr	_sifields._sigfault._addr
#define	lsi_band	_sifields._sigpoll._band
#define	lsi_fd		_sifields._sigpoll._fd

/*
 * We make l_rt_sigframe exactly the same as that of Linux.
 * This is a different choice of FreeBSD's Linuxulator.
 */

struct l_rt_sigframe {
  l_ucharptr_t sf_pretcode;
  struct l_ucontext	sf_sc;
  struct l_siginfo	sf_si;
  /* fp state should follow */
};

#endif
