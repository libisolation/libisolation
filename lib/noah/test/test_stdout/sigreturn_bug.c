#include <stdio.h>
#include <signal.h>
#include <pthread.h>
#include <stdint.h>
#include <unistd.h>
#include <sys/wait.h>
#include <errno.h>

static void handler(int sig) {
  if (sig == SIGCHLD)
    printf("sigchld arrived\n");
}

int main() {
  struct sigaction sa;
  sa.sa_handler = handler;
  sa.sa_flags = 0;
  sigemptyset(&sa.sa_mask);
  sigaction(SIGCHLD, &sa, NULL);

  int pid;
  if ((pid = fork()) > 0) {
    printf("child pid\n");
  again:
    errno = 0;
    int dead_pid = waitpid(-1, 0, 0);
    if (dead_pid < 0 && errno == EINTR) {
      goto again;
    }
    if (dead_pid < 0)
      perror("bug");
    else {
      if (dead_pid == pid)
        printf("dead pid equals pid!\n");
      else
        printf("bug\n");
    }
  } else {
    sleep(1);
    _exit(0);
  }
}
