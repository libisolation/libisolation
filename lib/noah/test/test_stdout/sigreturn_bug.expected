child pid
sigchld arrived
dead pid equals pid!
